### ISSUE TITLE
Slower performance of eq and fill_ compared to numpy

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#5436]
```shell
python -m timeit -n 1000 'import torch; torch.ByteTensor(500, 500, 3).fill_(1) == torch.ByteTensor([1, 2, 3])'
python -m timeit -n 1000 'import numpy; numpy.ones((500, 500, 3), dtype = numpy.uint8) == numpy.array([1, 2, 3], dtype = numpy.uint8)'
```

```
1000 loops, best of 3: 4.81 msec per loop
1000 loops, best of 3: 1.45 msec per loop
```

`torch.__version__ = '0.4.0a0+1fdb392'`
`numpy.__version__ = '1.14.0'`

**Comment by yf225** [comment]
@vadimkantorov I got the following result when running the perf tests on current master:

```
python -m timeit -n 1000 'import torch; torch.ByteTensor(500, 500, 3).fill_(1) == torch.ByteTensor([1, 2, 3])'
python -m timeit -n 1000 'import numpy; numpy.ones((500, 500, 3), dtype = numpy.uint8) == numpy.array([1, 2, 3], dtype = numpy.uint8)'
```

```
1000 loops, best of 3: 2.27 msec per loop
1000 loops, best of 3: 1.98 msec per loop
```

It's still slower compared to numpy, but the difference is smaller. Could you also try to run the perf comparison again to see if it has been improved?

**Comment by vadimkantorov** [comment]
@yf225 Due to some build issues I can't install a fresh version from source, so can't test the recent improvements.

The right hand side array is very small, I hope that proper vectorization is done (three bytes fit into an int, so vectorized comparisons with a broadcasted rhs should be possible).

**Comment by yf225** [comment]
@vadimkantorov Got it, what are the build issues you are having?

@cpuhrsch Do you have any insights regarding vectorization?

**Comment by cpuhrsch** [comment]
This is currently work in progress as part of [this pull request](https://github.com/pytorch/pytorch/pull/7655). It's definitely on our list. We're also adding a more extensive set of microbenchmarks [to our benchmark repository](https://github.com/pytorch/benchmark) to compare to NumPy.

**Comment by vadimkantorov** [comment]
Byte array of size 3 doesn't fit well into 4-byte-aligned vector registers, but maybe some efficient packing is still possible...

@yf225 It was something related to PyTorch thinking that MKL was required and failing to discover it (instead I had self-built OpenBLAS)

**Comment by cpuhrsch** [comment]
@vadimkantorov - May I also ask what the use-case is? Maybe there's a way in which we could make this more efficient by specializing on something else.

**Comment by vadimkantorov** [comment]
Ground truth for image segmentation (e.g. in pascal voc) is often distributed as regular images, with a set of pre-specified RGB-colors, each of which corresponds to a class label. So to recover class-label map I need to perform such comparisons.

Ideally, I'd like to compare to all colors of the palette at the same time, but not sure if it's achievable.

**Comment by yf225** [comment]
@vadimkantorov For the build issues, do you mind opening a new issue with the build command and the CMake logs you saw? We would actually want to fix this use case.

**Comment by vadimkantorov** [comment]
I reinstalled PyTorch in a fresh conda environment with MKL. I get:
```
1000 loops, best of 3: 1.92 msec per loop # pytorch
1000 loops, best of 3: 1.26 msec per loop # numpy
```

**Comment by ezyang** [comment]
On master

```
python -m timeit -n 1000 'import torch; torch.ByteTensor(500, 500, 3).fill_(1) == torch.ByteTensor([1, 2, 3])'
1000 loops, best of 5: 605 usec per loop

python -m timeit -n 1000 'import numpy; numpy.ones((500, 500, 3), dtype = numpy.uint8) == numpy.array([1, 2, 3], dtype = numpy.uint8)'
1000 loops, best of 3: 1.88 msec per loop
```

Mission accomplished, let's go home.

**Comment by vadimkantorov** [comment]
@ezyang A question: when the last dimension is small, does it do comparison with triplets of numbers in one go in a vectorized fashion? In theory it could even load the rhs in a vector register once, since broadcasting is dummy

**Comment by ezyang** [comment]
Not sure. But eq is going to be used with TensorIterator and it's going to coalesce contiguous dimensions, so my guess is that something reasonable will happen.

**Comment by vadimkantorov** [comment]
Small dimension ops (widh size = 2/3/4) may get more used with rise of point cloud processing in PyTorch. I think it makes sense to add more tests about these sizes.

**Comment by vadimkantorov** [comment]
Another issue would be fast converting [B x 3 x H x W] uint8 tensors to [B x H x W] uint32 tensor (with prepended zero byte)

**Comment by vadimkantorov** [comment]
> Not sure. But eq is going to be used with TensorIterator and it's going to coalesce contiguous dimensions, so my guess is that something reasonable will happen.

Will it be able to optimize it even when dimension is 3 and not 4 (or in general not an exact divisor of vector register, so probably masked vector ops are needed on CPU)

**Comment by ezyang** [comment]
> Will it be able to optimize it even when dimension is 3 and not 4 (or in general not an exact divisor of vector register, so probably masked vector ops are needed on CPU)

For equality, yeah. It won't divide nicely but it doesn't need to: we'll just do as much as divides neatly by the vector size and then use an iteration to fill things up in the end. The other situation you posted about above (convert triplets of uint8 into uint32) is probably harder because the data read-in and write-out isn't uniform.
